ntbs-service/DataMigration/NotificationMapper.cs

```diff
@@ -60,37 +65,77 @@ public NotificationMapper(IConfiguration _configuration)
+            }
+        }
 
-                return notifications;
+        public async Task<List<List<Notification>>> GetById(List<string> notificationIds)
+        {
+            using (var connection = new SqlConnection(connectionString))
+            {
+                connection.Open();
+
+                var whereClause = string.Format(WhereClauseById, string.Join(",", notificationIds.Select(n => $@"'{n}'")));
+                var queryWithWhereClause = string.Format(NotificationsQuery, whereClause);
+
+                var notificationsRaw = await connection.QueryAsync(queryWithWhereClause);
+                var legacyIds = notificationsRaw.Select(x => x.OldNotificationId).Cast<string>();
+
+                var notificationSitesRaw = await GetNotificationSites(connection, legacyIds);
+
+                return GetGroupedResultsAsNotification(notificationsRaw, notificationSitesRaw);
             }
         }
 
-        private static NotificationSite AsNotificationSite(dynamic result)
+        private async Task<IEnumerable<dynamic>> GetNotificationSites(SqlConnection connection, IEnumerable<string> legacyIds)
         {
-            return new NotificationSite {
-                SiteId = result.DiseaseSiteId,
-                SiteDescription = result.DiseaseSiteText
-            };
+            if (legacyIds.Count() == 0)
+            {
+                return new List<dynamic>();
+            }
+            var idsForInClause = string.Join(",", legacyIds.Select(x => $@"'{x}'"));
+            return await connection.QueryAsync(string.Format(NotificationSitesQuery, idsForInClause));
+        }
+
+        private List<List<Notification>> GetGroupedResultsAsNotification(IEnumerable<dynamic> notificationsRaw, IEnumerable<dynamic> notificationSitesRaw)
+        {
+            var groupedResults = new List<List<Notification>>();
+            var notificationGroups = notificationsRaw.GroupBy(x => x.GroupId);
+            var notificationSiteGroups = notificationSitesRaw.GroupBy(x => x.OldNotificationId);
+            
+            var notificationSiteDictionary = new Dictionary<string, List<NotificationSite>>();
+            foreach (var notificationSiteGroup in notificationSiteGroups)
+            {
+                notificationSiteDictionary.Add(notificationSiteGroup.Key, notificationSiteGroup.Select(AsNotificationSite).ToList());
+            }
+
+            foreach (var notificationGroup in notificationGroups)
+            {
+                var notifications = notificationGroup.Select(AsNotification).ToList();
+                notifications.ForEach(x => {
+                    if (notificationSiteDictionary.ContainsKey(x.LegacyId))
+                    {
+                        x.NotificationSites = notificationSiteDictionary[x.LegacyId];
+                    }
+                });
+                
+                groupedResults.Add(notifications);
+            }
+
+            return groupedResults;

```

Comment:
This whole method seems to actually be doing two different things and does them in a bit of a intertwined way. the things it does:
- attaches sites to notifications
- creates groups out of notifications based on patient

I feel like it could become much cleaner if it separated the two actions, particularly in that order - after all the second grouping does not depend on the first being done at all